File: openai-agents/src/ollie_integrations_openai_agents/signals.py
"""Behavior signals for OpenAI Agents single-run traces."""

from __future__ import annotations

import os
from typing import Any

SAFETY_FINISH = frozenset({"safety", "content_filter", "content_filtered"})
TRUNCATED_FINISH = frozenset({"length", "max_tokens", "max_output_tokens"})


def _span_type(span: dict[str, Any]) -> str:
    return str(span.get("type") or "").strip()


def _span_status(span: dict[str, Any]) -> str:
    return str(span.get("status") or "success").strip()


def _high_latency_ms() -> int:
    try:
        return max(1, int(os.getenv("OLLIE_HIGH_LATENCY_MS", "30000")))
    except ValueError:
        return 30000


def direct_signals(
    *,
    spans: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """Framework-direct behavior signals from span fields (P1 / guardrail)."""
    trigger: list[dict[str, Any]] = []
    context: list[dict[str, Any]] = []

    for span in spans:
        st = _span_type(span)
        status = _span_status(span)
        if st == "guardrail" and (status == "failure" or span.get("triggered")):
            trigger.append({"name": "guardrail_blocked", "guardrail": span.get("name")})
        elif st == "tool" and status == "failure":
            context.append({"name": "tool_error", "tool": span.get("name")})
        elif st == "llm" and status == "failure":
            context.append({"name": "llm_error", "llm": span.get("name")})
        elif st == "llm":
            fr = str(span.get("finish_reason") or "").strip().lower()
            if fr in TRUNCATED_FINISH:
                context.append({"name": "output_truncated"})
            elif fr in SAFETY_FINISH:
                context.append({"name": "safety_stop"})

    return trigger, _dedupe(context)


def derived_signals(
    *,
    spans: list[dict[str, Any]],
    output: str,
    success: bool,
    latency_ms: int,
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """P2 session-bad outcomes + repeated tool failure pattern."""
    trigger: list[dict[str, Any]] = []
    context: list[dict[str, Any]] = []

    if not success:
        context.append({"name": "runtime_failure"})
    if not str(output or "").strip():
        context.append({"name": "empty_final_response"})

    tool_spans = [s for s in spans if _span_type(s) == "tool"]
    failed_tools = [s for s in tool_spans if _span_status(s) == "failure"]
    if len(failed_tools) >= 2:
        trigger.append({"name": "repeated_tool_error"})
    elif len(failed_tools) == 1:
        context.append({"name": "tool_error", "tool": failed_tools[0].get("name")})

    if latency_ms >= _high_latency_ms():
        context.append({"name": "high_latency"})

    return trigger, _dedupe(context)


def instrument_events(
    *,
    spans: list[dict[str, Any]],
    output: str,
    success: bool,
    latency_ms: int,
    interaction_ref: str = "ix_0",
) -> tuple[dict[str, Any], list[dict[str, Any]]]:
    """Build events (spans only; empty trigger/context) and anchored _signal_hits."""
    from ollie_integrations_openai_agents.hits import hits_from_named_signals

    trig_d, ctx_d = direct_signals(spans=spans)
    trig_b, ctx_b = derived_signals(spans=spans, output=output, success=success, latency_ms=latency_ms)
    trigger = _dedupe(trig_d + trig_b)
    context = _dedupe(ctx_d + ctx_b)
    pending = hits_from_named_signals(
        trigger=trigger,
        context=context,
        spans=spans,
        interaction_ref=interaction_ref,
    )
    return (
        {
            "trigger": [],
            "context": [],
            "spans": spans,
        },
        pending,
    )


def _dedupe(signals: list[dict[str, Any]]) -> list[dict[str, Any]]:
    seen: set[str] = set()
    out: list[dict[str, Any]] = []
    for sig in signals:
        name = str(sig.get("name") or "")
        if not name or name in seen:
            continue
        seen.add(name)
        out.append(sig)
    return out

File: openai-agents-ts/src/signals.ts
import { hitsFromNamedSignals } from "./hits.js";
import type { EventsDict, SignalHit, SpanRecord } from "./types.js";

const SAFETY_FINISH = new Set(["safety", "content_filter", "content_filtered"]);
const TRUNCATED_FINISH = new Set(["length", "max_tokens", "max_output_tokens"]);

function dedupe(signals: Array<{ name: string; [k: string]: unknown }>) {
  const seen = new Set<string>();
  return signals.filter((s) => {
    if (seen.has(s.name)) return false;
    seen.add(s.name);
    return true;
  });
}

function directSignals(spans: SpanRecord[]) {
  const trigger: Array<{ name: string; [k: string]: unknown }> = [];
  const context: Array<{ name: string; [k: string]: unknown }> = [];

  for (const span of spans) {
    if (span.type === "handoff") {
      context.push({ name: "delegation" });
    } else if (span.type === "guardrail" && (span.status === "failure" || span.triggered)) {
      trigger.push({ name: "guardrail_blocked", guardrail: span.name });
    } else if (span.type === "tool" && span.status === "failure") {
      context.push({ name: "tool_error", tool: span.name });
    } else if (span.type === "llm" && span.status === "failure") {
      context.push({ name: "llm_error", llm: span.name });
    } else if (span.type === "llm") {
      const props =
        span.properties && typeof span.properties === "object" ? span.properties : {};
      const fr = String(span.finish_reason ?? props.finish_reason ?? "").toLowerCase();
      if (TRUNCATED_FINISH.has(fr)) context.push({ name: "output_truncated" });
      else if (SAFETY_FINISH.has(fr)) context.push({ name: "safety_stop" });
    }
  }

  return { trigger, context: dedupe(context) };
}

function derivedSignals(
  spans: SpanRecord[],
  output: string,
  success: boolean,
  latencyMs: number,
) {
  const trigger: Array<{ name: string; [k: string]: unknown }> = [];
  const context: Array<{ name: string; [k: string]: unknown }> = [];

  if (!success) context.push({ name: "runtime_failure" });
  if (!output.trim()) context.push({ name: "empty_final_response" });
  if (spans.some((s) => s.type === "tool")) context.push({ name: "used_tool" });
  const failedTools = spans.filter((s) => s.type === "tool" && s.status === "failure");
  if (failedTools.length >= 2) trigger.push({ name: "repeated_tool_error" });
  if (spans.filter((s) => s.type === "handoff").length >= 2)
    context.push({ name: "multi_handoff" });
  const threshold = Number(process.env.OLLIE_HIGH_LATENCY_MS ?? 30000);
  if (latencyMs >= threshold) context.push({ name: "high_latency" });

  return { trigger, context: dedupe(context) };
}

export function instrumentEvents(
  spans: SpanRecord[],
  output: string,
  success: boolean,
  latencyMs: number,
  interactionRef = "ix_0",
): { events: EventsDict; signalHits: SignalHit[] } {
  const { trigger: trigD, context: ctxD } = directSignals(spans);
  const { trigger: trigB, context: ctxB } = derivedSignals(spans, output, success, latencyMs);
  const trigger = dedupe([...trigD, ...trigB]);
  const context = dedupe([...ctxD, ...ctxB]);
  const signalHits = hitsFromNamedSignals({
    trigger,
    context,
    spans,
    interactionRef,
  });
  return {
    events: { trigger: [], context: [], spans },
    signalHits,
  };
}
